src/toil/provisioners/aws/awsProvisioner.py

```diff
@@ -445,15 +447,35 @@ def destroyCluster(self):
     def terminateNodes(self, nodes : List[Node]):
         self._terminateIDs([x.name for x in nodes])
 
-    def addNodes(self, nodeType, numNodes, preemptable, spotBid=None) -> int:
+    def addNodes(self, nodeTypes: Set[str], numNodes, preemptable, spotBid=None) -> int:
         assert self._leaderPrivateIP
-        if preemptable and not spotBid:
-            if self._spotBidsMap and nodeType in self._spotBidsMap:
-                spotBid = self._spotBidsMap[nodeType]
+
+        if preemptable and spotBid is None:
+            if self._spotBidsMap and frozenset(nodeTypes) in self._spotBidsMap:
+                spotBid = self._spotBidsMap[frozenset(nodeTypes)]
+            elif len(nodeTypes) == 1:
+                # The Toil autoscaler forgets the equivalence classes. Find
+                # some plausible equivalence class.
+                instance_type = next(iter(nodeTypes))
+                for types, bid in self._spotBidsMap.items():
+                    if instance_type in types:
+                        # We bid on a class that includes this type
+                        spotBid = bid
+                        break
+                if spotBid is None:
+                    # We didn't bid on any class including this type either
+                    raise RuntimeError("No spot bid given for a preemptable node request.")
             else:
                 raise RuntimeError("No spot bid given for a preemptable node request.")
-        instanceType = E2Instances[nodeType]
-        bdm = self._getBoto2BlockDeviceMapping(instanceType, rootVolSize=self._nodeStorageOverrides.get(nodeType, self._nodeStorage))
+
+        # We don't support any balancing here so just pick one of the
+        # equivalent node types
+        node_type = next(iter(nodeTypes))
```

Comment:
This chooses an arbitrary instance type from the set the user said were equivalent for them. I think it's up to how Python implements `set` whether it's always going to be the same one.

Since we aren't using an AWS ASG here, there's no way to actually balance across the spot pools. Since the set of instance types are supposed to be all the same to the user, we can just pick one arbitrarily. If that creates problems, we could move to a real random choice and spray across all the options.

We *could* write balancing code for Toil-workflow-managed Mesos autoscaling, but since I think it is going away I don't think it would be worth the effort. We want to use the new Kubernetes-based scaling specifically because it is smart enough to do stuff like that for us.